test/functional/api/keymap_spec.lua

```diff
@@ -0,0 +1,243 @@
+    -- The table will be the same except for the mode
+    local insert_table = local_copy(foo_bar_map_table)
+    insert_table['mode'] = 'i'
+    insert_table['nowait'] = 1
+
+    eq({insert_table}, nvim('get_keymap', 'i'))
+  end)
+
+  it('considers scope', function()
+    -- change the map slightly and add a <nowait> option to the map
+    nvim('command', 'nnoremap foo_longer bar_longer')
+    local foolong_bar_map_table = local_copy(foo_bar_map_table)
+    foolong_bar_map_table['lhs'] = 'foo_longer'
+    foolong_bar_map_table['rhs'] = 'bar_longer'
+
+    local foo_bar_buffer_string = 'nnoremap <buffer> foo bar'
+    local buffer_table = local_copy(foo_bar_map_table)
+    buffer_table['buffer'] = 1
+
+    nvim('command', foo_bar_buffer_string)
+
+    -- The buffer mapping should not show up
+    eq({foolong_bar_map_table}, nvim('get_keymap', 'n'))
+    eq({buffer_table}, nvim('buf_get_keymap', 0, 'n'))
+  end)
+
+  it('considers scope for overlapping maps', function()
+    nvim('command', 'nnoremap foo bar')
+
+    local buffer_table = local_copy(foo_bar_map_table)
+    buffer_table['buffer'] = 1
+
+    nvim('command', 'nnoremap <buffer> foo bar')
+
+    eq({foo_bar_map_table}, nvim('get_keymap', 'n'))
+    eq({buffer_table}, nvim('buf_get_keymap', 0, 'n'))
+  end)
+
+  -- Test toggle switches for basic options
+  -- @param  option  The key represented in the `maparg()` result dict
+  local function global_and_buffer_test(map,
+                                        option,
+                                        option_token,
+                                        global_on_result,
+                                        buffer_on_result,
+                                        global_off_result,
+                                        buffer_off_result,
+                                        new_windows)
+
+    local function make_new_windows(number_of_windows)
+      if new_windows == nil then
+        return nil
+      end
+
+      for _=1,number_of_windows do
+        nvim('command', 'new')
+      end
+    end
+
+    local mode = string.sub(map, 1,1)
+    -- Don't run this for the <buffer> mapping, since it doesn't make sense
+    if option_token ~= '<buffer>' then
+      it(string.format( 'returns %d for the key "%s" when %s is used globally with %s (%s)',
+          global_on_result, option, option_token, map, mode), function()
+        make_new_windows(new_windows)
+        nvim('command', map .. ' ' .. option_token .. ' foo bar')
+        local result = nvim('get_keymap', mode)[1][option]
+        eq(global_on_result, result)
+      end)
+    end
+
+    it(string.format('returns %d for the key "%s" when %s is used for buffers with %s (%s)',
+        buffer_on_result, option, option_token, map, mode), function()
+      make_new_windows(new_windows)
+      nvim('command', map .. ' <buffer> ' .. option_token .. ' foo bar')
+      local result = nvim('buf_get_keymap', 0, mode)[1][option]
```

Comment:
You are not checking that nvim_buf_get_keymap is working for non-current buffer anywhere.

Additionally I would recommend throwing away `nvim`, there is `[helpers.]*meths`: `meths.get_keymap(mode)`, `curbufmeths.get_keymap(mode)` and `bufmeths.get_keymap(buf, mode)`, `[helpers.]command` for `meths.command`: this variant is more convenient. Should be possible to translate just by executing (in order present)

```Vim
:%s/nvim('buf_\(\w\+\)', 0, /curbufmeths.\1(/g
:%s/nvim('buf_\(\w\+\)', /bufmeths.\1(/g
:%s/nvim('command', /command(/g
:%s/nvim('\(\w\+\)', /meths.\1/g
```